File: Stack/leetcode_20.cpp
// Given a string s containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.
// An input string is valid if : Open brackets must be closed by the same type of brackets.Open brackets must 
// be closed in the correct order.Every close bracket has a corresponding open bracket of the same type.

#include <iostream>
#include <unordered_map>
#include <stack>

    using namespace std;

class Solution
{
public:
    bool isValid(string s)
    {
        unordered_map<char, char> closing_brackets = {{')', '('}, {'}', '{'}, {']', '['}};
        stack<char> stk;

        for (char c : s)
        {
            if (closing_brackets.count(c))
            {
                if (stk.empty() || stk.top() != closing_brackets[c])
                {
                    return false;
                }
                stk.pop();
            }
            else
            {
                stk.push(c);
            }
        }

        return stk.empty();
    }
};

File: Stack/leetcode_20.py
# Given a string s containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.
# An input string is valid if : Open brackets must be closed by the same type of brackets.Open brackets must
# be closed in the correct order.Every close bracket has a corresponding open bracket of the same type.

class Solution:
    def isValid(self, s: str) -> bool:
        """
        Checks if a string contains valid parentheses.

        Args:
            s: A string.

        Returns:
            True if the string contains valid parentheses, False otherwise.
        """

        stack = []
        for c in s:
            if c in "([{":
                stack.append(c)
            elif c in ")]}":
                if not stack or stack.pop() != c:
                    return False

        return len(stack) == 0
